### PR TITLE
[AArch64] Add support for Cortex-R82AE and improve Cortex-R82

### PR DESCRIPTION
**Pull request by jthackray**
- [AArch64] Add support for Cortex-R82AE and improve Cortex-R82

### CODE REVIEW

**Review by Stylie777 — APPROVED** [review]

File: llvm/lib/Target/AArch64/AArch64Subtarget.cpp [hunk 1]
```diff
@@ -143,6 +143,7 @@ void AArch64Subtarget::initializeProperties(bool HasMinSize) {
   case CortexA78AE:
   case CortexA78C:
   case CortexR82:
+  case CortexR82AE:
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64Subtarget.cpp:146]: Can you move both of these into the block with CortexA55? It has always been in the wrong place compared to the other cpus around it and the alignments it sets.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 2]
```diff
@@ -632,7 +632,18 @@ inline constexpr CpuInfo CpuInfos[] = {
                                AArch64::AEK_PAUTH, AArch64::AEK_SVE2BITPERM,
                                AArch64::AEK_FLAGM, AArch64::AEK_PERFMON,
                                AArch64::AEK_PREDRES, AArch64::AEK_PROFILE})},
-    {"cortex-r82", ARMV8R, AArch64::ExtensionBitset({AArch64::AEK_LSE})},
+    {"cortex-r82", ARMV8R,
+     AArch64::ExtensionBitset({AArch64::AEK_CRC, AArch64::AEK_DOTPROD,
```
Comment (davemgreen) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:636]: Are these implied by ARMV8R? Is there an advantage to having them specified separately too?

File: llvm/lib/Target/AArch64/AArch64Subtarget.cpp (hunk 1 above)
Comment (jthackray) [reply llvm/lib/Target/AArch64/AArch64Subtarget.cpp]: Thanks, done.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 2 above)
Comment (jthackray) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h]: Thanks, good spot, yes these can be trimmed.

- AEK_FP is enabled for ARMV8A
- AEK_LSE is enabled for ARMV8_1A
- AEK_CRC is enabled for ARMV8_1A
- AEK_RDM is enabled for ARMV8_1A
- AEK_RAS is enabled for ARMV8_2A
- AEK_RCPC is enabled for ARMV8_3A
- AEK_PAUTH is enabled for ARMV8_3A
- AEK_DOTPROD is enabled for ARMV8_4A
- AEK_FLAGM doesn't default
- AEK_PERFMON doesn't default
- AEK_PREDRES doesn't default

ARMV8R inherits from ARMV8_5A.DefaultExts, so we only need the latter 3 flags.

Code now updated.

Comment (davemgreen) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h]: Can you check if ARMV8R enables AEK_LSE? There is a flip(), which might be turning it off?

Comment (jthackray) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h]: Ah yes. Now definitely enabled in the latest commit.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Pull request merged**